### PR TITLE
refine repository and config loader

### PR DESCRIPTION
**Pull request by OsmondJiang**
1. use repository provider to get repository, it's a wrapper of RestoreGitMap
2. config loader to use Input to load config
3. Repository provider is part of Input
4. hide current/fallback docset to some of the components

The component dependencies:

`->` means `Depends`

1. RepositoryProvider -> Docset Path/Config/RestoreGitMap/LocalizationRepository
2. Input -> RepositoryProvider
3. ConfigLoader -> Input
4. Config -> ConfigLoader

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5118)

### CODE REVIEW

File: src/docfx/config/ConfigLoader.cs [hunk 1]
```diff
@@ -66,24 +80,39 @@ private static (List<Error>, Config) LoadCore(string docsetPath, CommandLineOpti
...
             {
                 var extendErrors = new List<Error>();
-                (extendErrors, configObject) = ExtendConfigs(configObject, docsetPath);
+                (extendErrors, configObject) = ExtendConfigs(configObject);
                 errors.AddRange(extendErrors);
             }
 
             // apply overwrite
-            OverwriteConfig(configObject, locale ?? options.Locale, GetBranch());
+            OverwriteConfig(configObject, locale ?? options.Locale, _repositoryProvider.GetRepository(FileOrigin.Default)?.Branch);
```
Comment (live1206) [inline src/docfx/config/ConfigLoader.cs]: If only branch needed here, consider not depend on `_repositoryProvider`?

File: src/docfx/build/Build.cs [hunk 2]
```diff
@@ -14,46 +13,59 @@ internal static class Build
...
-                    docsetPath, locale, config, repository, fallbackRepo, restoreGitMap);
-                var outputPath = Path.Combine(docsetPath, config.Output.Path);
-                var dependencyDocsets = LoadDependencies(docset, restoreGitMap);
+                // get docsets(build docset, fallback docset and dependency docsets)
+                repositoryProvider.Config(config);
+                var (docset, fallbackDocset) = GetDocsetWithFallback(docsetPath, locale, config, repositoryProvider, restoreGitMap);
+                if (!string.Equals(docset.DocsetPath, PathUtility.NormalizeFolder(docsetPath), PathUtility.PathComparison))
+                {
+                    // entry docset is not the docset to build
+                    input = new Input(docset.DocsetPath, repositoryProvider);
```
Comment (yufeih) [inline src/docfx/build/Build.cs]: Is there a way to detect if we are building from loc or building from English upfront?

File: src/docfx/docset/RepositoryProvider.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                        {
+                            // point to a folder
+                            return (dependencyPath, null);
+                        }
+
+                        return (dependencyPath, Repository.Create(dependencyPath, dependency.Branch, dependency.Url, dependencyCommit));
+                    })).Value;
+            }
+
+            return default;
```
Comment (yufeih) [inline src/docfx/docset/RepositoryProvider.cs]:
```suggestion
            throw new InvalidOperationException();
```

File: src/docfx/docset/RepositoryProvider.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+
+namespace Microsoft.Docs.Build
+{
+    internal class RepositoryProvider
+    {
+        private readonly ConcurrentDictionary<string, Lazy<(string docset, Repository repository)>> _repositories = new ConcurrentDictionary<string, Lazy<(string docset, Repository repository)>>();
```
Comment (yufeih) [inline src/docfx/docset/RepositoryProvider.cs]:
```suggestion
        private readonly ConcurrentDictionary<(FileOrigin, string dependencyName), Lazy<(string docset, Repository repository)>> _repositories = new ConcurrentDictionary<string, Lazy<(string docset, Repository repository)>>();
```

File: src/docfx/build/Build.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/build/Build.cs]: we don't until we get the configuration

File: src/docfx/docset/RepositoryProvider.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        private static Repository GetFallbackRepository(
+            Repository repository,
+            RestoreGitMap restoreGitMap)
+        {
+            Debug.Assert(restoreGitMap != null);
```
Comment (live1206) [inline src/docfx/docset/RepositoryProvider.cs]: put this assertion in `ConfigRestoreMap` instead?

File: src/docfx/lib/path/PathUtility.cs [hunk 6]
```diff
@@ -60,13 +60,13 @@ public static (bool match, bool isFileMatch, string remainingPath) Match(this st
         /// </summary>
         public static string FindYamlOrJson(string pathWithoutExtension)
         {
-            var fullPath = PathUtility.NormalizeFile(pathWithoutExtension + ".yml");
+            var fullPath = NormalizeFile(pathWithoutExtension + ".yml");
```
Comment (live1206) [inline src/docfx/lib/path/PathUtility.cs]: does this mean the config file could be from .git?

**Pull request merged**